python/shotgun_data/shotgun_data_retriever.py

```diff
@@ -0,0 +1,723 @@
+    _SG_UPDATE_PRIORITY = 60
+    _SG_FIND_PRIORITY = 70
+    _CHECK_THUMB_PRIORITY = 80
+
+    # ------------------------------------------------------------------------------------------------
+    # Signals
+    #
+    # syntax: work_completed(uid, request_type, data_dict)
+    # - uid is a unique id which matches the unique id
+    #   returned by the corresponding request call.
+    #
+    # - request_type is a string denoting the type of request
+    #   this event is associated with. It can be either "find"
+    #   or "thumbnail"
+    #
+    # - data_dict is a dictionary containing the payload
+    #   of the request. It will be different depending on
+    #   what type of request it is.
+    #
+    #   For find() requests, the data_dict will be on the form
+    #   {"sg": data }, where data is the data returned by the sg API
+    #
+    #   For thumbnail requests, the data dict will be on the form
+    #   {"thumb_path": path}, where path is a path to a location
+    #   on disk where the thumbnail can be accessed.
+    work_completed = QtCore.Signal(str, str, dict)
+
+    # syntax: work_failure(uid, error_message)
+    # - uid is a unique id which matches the unique id
+    #   returned by the corresponding request call.
+    # - error message is an error message string.
+    work_failure = QtCore.Signal(str, str)
+
+    def __init__(self, parent=None, sg=None, bg_task_manager=None):
+        """
+        Construction
+        """
+        QtCore.QObject.__init__(self, parent)
+        self._bundle = sgtk.platform.current_bundle()
+
+        # set up the background task manager:
+        self._task_manager = bg_task_manager or BackgroundTaskManager(parent=self, max_threads=1)
+        self._owns_task_manager = (bg_task_manager is None)
+        self._bg_tasks_group = self._task_manager.next_group_id()
+        self._task_manager.task_completed.connect(self._on_task_completed)
+        self._task_manager.task_failed.connect(self._on_task_failed)
+
+        self._thumb_task_id_map = {}
+
+    # ------------------------------------------------------------------------------------------------
+    # Public interface
+
+    @property
+    def _shotgun_connection(self):
+        """
+        Returns the Shotgun connection currently being used.  Note that this will be thread
+        specific.
+
+        :returns:    The Shotgun connection for this instance in the current thread
+        """
+        return self._task_manager.shotgun_connection
+
+    def start(self):
+        """
+        Start the retriever thread
+        """
+        self._task_manager.start_processing()
+
+    def stop(self):
+        """
+        Stop the retriever thread
+        """
+        if not self._task_manager:
+            return
+
+        if self._owns_task_manager:
+            # we own the task manager so we'll need to completely shut it down before
+            # returning
+            self._task_manager.shut_down()
+            self._task_manager = None
```

Comment:
A comment in the method (or maybe if the class docstring?) mentioning the data retriever can't be restarted would be nice, since it's behaviour is contrary to the BgTaskManager.